### PR TITLE
SEG-19

### PR DESCRIPTION
**Pull request by vjnathe-webonise**
Add Single Page App (SPA) flag to VWO Destination

### CODE REVIEW

**Comment by vjnathe-webonise** [comment]
@CarlosMecha @ameyawebonise This PR is derived from `SEG-18` so also has code from it to avoid the version conflicting. Once `SEG-18` merged to master this will only display changes with respect to `SEG-19`.

File: integrations/visual-website-optimizer/HISTORY.md [hunk 1]
```diff
@@ -1,3 +1,14 @@
+2.5.0 / 2019-10-03
```
Comment (CarlosMecha) [inline integrations/visual-website-optimizer/HISTORY.md]: Let's do only one version bump (like #12) and release both changes at the same time.

File: integrations/visual-website-optimizer/lib/index.js [hunk 2]
```diff
@@ -24,7 +24,8 @@ var VWO = (module.exports = integration('Visual Website Optimizer')
   .option('useExistingJQuery', false)
   .option('replay', true)
   .option('listen', false)
-  .option('experimentNonInteraction', false));
+  .option('experimentNonInteraction', false)
+  .option('isSPA', false));
```
Comment (CarlosMecha) [inline integrations/visual-website-optimizer/lib/index.js]: I like the idea of letting the customers choose if they want to send events for all experiments. I'm aware this wasn't on the initial requirements, but it should be easy to add: Let's add a new setting `trackOnlyABExperiments` with default value `false`, as:
```javascript
.option('trackOnlyABExperiments'));
```

File: integrations/visual-website-optimizer/lib/index.js [hunk 3]
```diff
@@ -182,6 +183,19 @@ function enqueue(fn) {
...
 
+/**
+ * Check whether the experiment type is one of allowed types.
+ *
+ * @param {Object} experiment
+ * @return {Boolean}
+ */
+function isValidExperimentType(experiment) {
+  var allowedExperimentTypes = ['VISUAL_AB', 'VISUAL', 'SPLIT_URL', 'SURVEY'];
+  var experimentType = (experiment && experiment.type) || '';
```
Comment (CarlosMecha) [inline integrations/visual-website-optimizer/lib/index.js]: To make it easier to read, and to check the new setting, let's do:
```javascript
function isValidExperimentType(experiment, options) {
  if (!Boolean(this.options.trackOnlyABExperiments)) {
    return true;
  }
  if (!experiment || !experiment.type) {
    return false;
  }
  return allowedExperimentTypes.indexOf(experimentType.toUpperCase()) !== -1;
}
```
Then we can add the `allowedExperimentsTypes` array at the top of the file, so it's only created once.

File: integrations/visual-website-optimizer/test/index.test.js [hunk 4]
```diff
@@ -209,6 +214,35 @@ describe('Visual Website Optimizer', function() {
         done();
       });
     });
+
+    it('should not send experiments if experiment type is not expected', function(done) {
```
Comment (CarlosMecha) [inline integrations/visual-website-optimizer/test/index.test.js]: Let's also add a new test when the new setting is enabled or disabled

Comment (CarlosMecha) [reply integrations/visual-website-optimizer/test/index.test.js]: Do you think we can create a test for isSPA where we check the `<head>` tag and look for the url with the script including `is_spa=1`? Like described here: https://help.vwo.com/hc/en-us/articles/360020919254

Not sure how doable is it.

**Review by CarlosMecha — CHANGES_REQUESTED** [review]
Some minor changes. :rocket:

**Review by CarlosMecha — APPROVED** [review]
:boom:

File: integrations/visual-website-optimizer/test/index.test.js (hunk 4 above)
Comment (vjnathe-webonise) [reply integrations/visual-website-optimizer/test/index.test.js]: for testing `is_spa=1` inside head tag we might need to introduce enzyme for shallow rendering but doing this we might need some UI component to be in exist to load that which i think not possible.

Comment (CarlosMecha) [reply integrations/visual-website-optimizer/test/index.test.js]: Agree.

**Comment by vjnathe-webonise** [comment]
Closing this PR as it's merged to upstream.

**Pull request closed without merging**